std/assembly/typedarray.ts

```diff
@@ -1718,3 +1762,70 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
   return changetype<TArray>(out); // retains
 }
+
+// @ts-ignore: decorator
+@inline
+function SET<TArray extends ArrayBufferView, T, UArray extends ArrayBufferView, U>(target: TArray, source: UArray, offset: i32 = 0): void {
+  // need to assert at compile time that U is not a reference or a function
+  if (isReference<U>()) {
+    ERROR(E_NOTIMPLEMENTED);
+  }
+
+  // Uncaught RangeError: offset is out of bounds
+  if (offset < 0) throw new RangeError(E_INDEXOUTOFRANGE);
+  if (source.length + offset > target.length) throw new RangeError(E_INDEXOUTOFRANGE);
+
+  // if the types align and match, use memory.copy() instead of manual loop
+  if (isInteger<T>() == isInteger<U>() && alignof<T>() == alignof<U>()) {
+    memory.copy(
+      target.dataStart + (<usize>offset << alignof<T>()),
+      source.dataStart,
+      source.byteLength
+    );
+  } else {
+    let targetDataStart = target.dataStart + (<usize>offset << alignof<T>());
+    let sourceDataStart = source.dataStart;
+    let count = source.length;
+    for (let i = 0; i < count; i++) {
+      // if TArray is Uint8ClampedArray, then values must be clamped
+      if (target instanceof Uint8ClampedArray) {
+        if (isFloat<U>()) {
+          let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
+          store<T>(
+            targetDataStart + (<usize>i << alignof<T>()),
+            isFinite<U>(value) ? <T>max<U>(0, min<U>(255, value)) : 0
+          );
+        } else {
+          let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
+          if (!isSigned<U>()) {
+            store<T>(
+              targetDataStart + (<usize>i << alignof<T>()),
+              // @ts-ignore: cast to T is valid for numeric types here
+              min<U>(255, value)
+            );
+          } else if (sizeof<T>() <= 4) {
+            store<T>(
+              targetDataStart + (<usize>i << alignof<T>()),
+              // @ts-ignore: cast to T is valid for numeric types here
+              ~(<i32>value >> 31) & (((255 - <i32>value) >> 31) | value)
+            );
+          } else {
+            store<T>(
+              targetDataStart + (<usize>i << alignof<T>()),
+              // @ts-ignore: cast to T is valid for numeric types here
+              ~(<i64>value >> 63) & (((255 - <i64>value) >> 63) | value)
+            );
+          }
+        }
+        // if U is a float, then casting float to int must include a finite check
+      } else if (isFloat<U>() && !isFloat<T>()) {
+        let value = load<U>(sourceDataStart + (<usize>i << alignof<U>()));
+        // @ts-ignore: cast to T is valid for numeric types here
+        store<T>(targetDataStart + (<usize>i << alignof<T>()), isFinite<U>(value) ? <T>value : 0);
+      } else {
+        // @ts-ignore: cast to T is valid for numeric types here
+        store<T>(targetDataStart + (<usize>i << alignof<T>()), <T>load<U>(sourceDataStart + (<usize>i << alignof<U>())));
```

Comment:
Looks like this is the int->int case, in which case the `<T>` is unnecessary for the same reason. I believe the tests currently do not catch this due to missing small and long int cases.

Reply:
```
trace("hit!");
store<T>(targetDataStart + (<usize>i << alignof<T>()), <T>load<U>(sourceDataStart + (<usize>i << alignof<U>())));
```

This is the line in question yes?

```
- instantiate untouched OK (33.971 ms)
  716 allocs, 716 frees, 2077 increments, 2077 decrements

  [fun] __start = function 674() { [native code] }
  [obj] memory = [object WebAssembly.Memory]

  [start]
  trace: Hit!
  trace: Hit!
  trace: Hit!
...
```

This line is hit quite regularly. I will simply remove the offending `<T>` and see what happens.

Reply:
Is it being hit with small int types for both T and U?

```ts
if (sizeof<U>() < 4) trace("hit U!");
if (sizeof<Z>() < 4) trace("hit Z!");
```

Reply:
Okay. I have evidence that this optimization isn't too necessary, but it does seem to help.

![image](https://user-images.githubusercontent.com/3761339/71118027-46d66e80-21a5-11ea-8b39-04fb70acbf34.png)

This is only in the `untouched` version, and the optimized version remains unchanged.

I had to change it so that the cast happens if the conversion is from integer to float, however, by adding another `else if` branch to take advantage of this optimization.

Reply:
The `i64` to native `i32` (or smaller) and vice-versa cases are interesting for the optimizer because it can potentially remain on i32 or i64 exclusively, without a `.wrap_i64` instruction for example.

Other interesting cases are `i8` -> `i32`, `u8` -> `i16`, `u16` -> `i8` because these involve sign extension/wrapping. That's what I'm trying to get to, as these aren't currently present in the tests.

Reply:
Okay. I'll talk with @MaxGraey to get the test cases all covered.